File: 62-unique-paths/unique-paths.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 */
var uniquePaths = function (m, n) {
    // right or down
    let memo = new Array(m).fill().map(() => new Array(n).fill(0));
    memo[0][0] = 1;
    // buttom-up
    for (let i = 0; i < m; i++) {
        for(let j = 0; j < n; j++){
            if(i > 0){
                memo[i][j] += memo[i-1][j]
            }
            if(j > 0){
                memo[i][j] += memo[i][j-1];
            }
        }
    }
    return memo[m-1][n-1];
};

File: 62-unique-paths/unique-paths.java
class Solution {
    public int uniquePaths(int m, int n) {

        int[][] visit = new int[m][n];

        // visit[0][0] = 1;
        // visit[0][1] = 1;
        // visit[1][0] = 1;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 || j == 0)
                    visit[i][j] = 1;
                else
                    visit[i][j] = visit[i - 1][j] + visit[i][j - 1];
            }
        }

        return visit[m - 1][n - 1];
    }
}